packages/financial-templates-lib/src/price-feed/CryptoWatchPriceFeed.ts

```diff
@@ -109,65 +110,72 @@ export class CryptoWatchPriceFeed extends PriceFeedInterface {
     // If the time is before the first piece of data in the set, return null because
     // the price is before the lookback window.
     if (time < firstPricePeriod.openTime - this.historicalTimestampBuffer) {
-      throw new Error(`${this.uuid}: time ${time} is before firstPricePeriod.openTime`);
+      throw new Error(`${this.uuid}: time ${time} is before firstPricePeriod.openTime minus historicalTimestampBuffer`);
     }
 
     // historicalPricePeriods are ordered from oldest to newest.
     // This finds the first pricePeriod whose closeTime is after the provided time.
     const match = this.historicalPricePeriods.find((pricePeriod) => {
-      return (
-        time < pricePeriod.closeTime + this.historicalTimestampBuffer &&
-        time >= pricePeriod.openTime - this.historicalTimestampBuffer
-      );
+      return time <= pricePeriod.closeTime && time >= pricePeriod.openTime;
     });
 
+    // If match doesn't succeed, then we can still try to find the nearest price and use it if its within the caller's
+    // allowed margin of error.
+    let returnPrice;
     if (match === undefined) {
-      // If match doesn't succeed, then give caller details about closest price periods.
+      // Accounting for the buffer, find the price periods before and after the target time. These will be undefined if
+      // not found. Reverse the search for the period so that we match the latest period that encompasses the target
+      // time.
       const before = this.historicalPricePeriods
         .slice()
         .reverse()
-        .find((pricePeriod) => time >= pricePeriod.openTime - this.historicalTimestampBuffer);
+        .find(
+          (pricePeriod) =>
+            time <= pricePeriod.closeTime + this.historicalTimestampBuffer && time >= pricePeriod.openTime
```

Comment:
So this means the result time must _strictly_ be before the requested time, right?

Reply:
Yes exactly, if its not `before`, then it could match as the `after` time. Do you think this is a problem?